### PR TITLE
Fix/color

### LINKED ISSUE #120: Document or Fix class ID sort order requirement
A common pattern is to ship only the annotations and retain the subtask config in the frontend wrapping ULabel. In doing so, I found a (usually correct) assumption about the inputs in the subtask.

If the class IDs in a subtask configuration are supplied out of order, ULabel shows incorrect coloring for the annotation. 

ULabel should either enforce ordering of class IDs at parse time or reorder as needed. Order shouldn't be assumed when supplying a class ID in the array of class configs.

### PR DESCRIPTION
**Pull request by csolbs24**
# Update annotation coloring

## Description

Closes #120 

Fixes annotations needing to be declared in class-id order in subtask definition.
Re-writes RecolorActiveItem ToolboxItem to fix issue with needing to re-initialize subtasks on color change, also makes code much more readable.
Fix issue with change id color pie not being updated properly when an annotation's color is updated.

## PR Checklist

- [x] Merged latest main
- [x] Version number in `package.json` has been bumped since last release
- [x] Version numbers match between package `package.json` and `src/version.js`
- [x] Ran `npm install` and `npm run build` AFTER bumping the version number
- [x] Updated documentation if necessary (currently just in `api_spec.md`)
- [x] Added changes to `changelog.md`

## Breaking API Changes

None

### CODE REVIEW

File: src/html_builder.ts [hunk 1]
```diff
@@ -15,486 +15,499 @@ import {
...
+    
+    style.appendChild(document.createTextNode(get_init_style(ulabel.config["container_id"])));
+}
 
-        return `<div class="mode-opt">
-            <a${href} id="md-btn--${md_key}" class="md-btn${sel}${subtask_classes} invert-this-svg" amdname="${md_name}">
-                ${svg_blob}
-            </a>
-        </div>`;
+function get_md_button(md_key, md_name, svg_blob, cur_md, subtasks) {
```
Comment (TrevorBurgoyne) [inline src/html_builder.ts:28]: Im dumb, what is `md`? mode? Maybe add some brief function descriptions

File: src/html_builder.ts [hunk 2]
```diff
@@ -15,486 +15,499 @@ import {
...
+        return false;
+    }
 
-        // Check if initial_crop exists and has the appropriate properties
-        let check_initial_crop: Function = function(initial_crop) {
-
-            // If initial_crop doesn't exist, return false
-            if (initial_crop == null) return false;
+    // Check if initial_crop exists and has the appropriate properties
+    let check_initial_crop: Function = function(initial_crop) {
```
Comment (TrevorBurgoyne) [inline src/html_builder.ts:188]: A name like `initial_crop_is_valid` or smth would read slightly better, for example
```js
if (initial_crop_is_valid(ulabel.config.initial_crop))
```
i get that this is just a refactor of the og function so just an idea to consider

File: src/html_builder.ts [hunk 3]
```diff
@@ -15,486 +15,499 @@ import {
...
         }
-        dialog_html += `
-            </svg>
-            <div class="centcirc"></div>
+        toolbox_html += `
         </div>`;
 
-        return dialog_html;
+        // Add dialog to the document
+        front_subtask_dialog_container_jq.append(front_dialog_html_v2); // TODO(new3d) MOVE THIS TO GLOB BOX -- superimpose atop thee anchor already there when needed, no remove and add back
```
Comment (TrevorBurgoyne) [inline src/html_builder.ts:386]: Is this comment still relavent?
```suggestion
        front_subtask_dialog_container_jq.append(front_dialog_html_v2);
```

File: src/html_builder.ts [hunk 4]
```diff
@@ -15,486 +15,499 @@ import {
...
...N52aK1kJSolqJk5JFfjzvSGhVSlI5bqd8uXrc6b7LusWFFaYIpebhG6Yo8yMscUOwRvL9O7YpwbWGKijCCpopAgmaKUIImivI+euLn6N+5vGDhUz9YghS9FOWCMz8TpMylvf98inLB5naNqFPZ3p/vHjX+Nb67WJqixSwLlllp9zXhpLYZydCFTdGZYBP4u5XhticWTbqKfaeoLuWLleF36a6UVtFhgmma/bUy/Js5rOU0DMapoFeGPylWTgX9MkxJ1XdjYIZfhvRu5cvxIT0zLN8Sx0f0zTDNkr3D5flwRL8Msy+7kUCiQ/plSIcWBb+W/gfXwyR5DPaepjod1mWK5beVodP70qo9bpjPFlX3wO6eD3O758OVu+fDij2yq2f8wvYZf1U4esbnpvfJU8T8nqbi/3ZY37UJ5y+G9H2pIEEKWIq6CVKgFHsEJQlSgBTNBIEUTQVD+B3wgGCPIsjv8QcF0fdiKAhi7KeRzERXE0TeE6UoKNnXlvq/r01ZEHVvotZJ5v/+Uk5RJ0GK/3uEd+zccF1BhH3eTIr6ggh79Tspmggi9Fv8pqi3yLT43zOz29TmCVIeD31P/go2it+078niC8yL9a59v7vqIJ0v3v146OH7D326RXIB30Nq3FLnKfzN/M3YJbkl/F7uaIhPNMQnGuITDfGJhvhEQ3yiIT7REJ9oiE80xCca4hMN8YmG+ERDfKIhPtEQn2iISfDv5Q7+3eqnAapHRanhT9+Ef/tXB2kHqB4UZYa/jSF+bvDsoTsClzxJDTudL2ApsiNwmxTFhkxrD1SKZ0OMaYqidyM8sR8CpciMof5Jke/YXXLNWTnKisoLNpcD7hPRZyAn6mQt67oaJl8j3OhYDUuho0i8Z1FbGNbSDl6PeLcZijCzmzlxHeTtnQp41agqxWKkj3lbwXW5lfQ/DnJj+K6R6yPqX1QR1Bj9PzZGimavUhkL6WR3OepvNvAD7RSxEqRoKuIJJkmho4i0yLRoXDRwLhMsyiliJkhRTBE1QYpSirgJUhRWVMRVtMvgpR/tQs8zkCL2KXqkVxE/QUrPcqPzIjGfkV40wkiQIkkxlAQpwhTDSZAiGMwUUoIUbkUNK0HKWYqhJUhhFEMUZG7gwjtFj/ymGGaClJ8UQ02QsiBZmpm/KByB+T7bX3ko8T9Zz1H5wFZx8QAAAABJRU5ErkJggg==">
```
Comment (TrevorBurgoyne) [inline src/html_builder.ts:441]: bruh

File: src/toolbox.ts [hunk 5]
```diff
@@ -1294,82 +1327,130 @@ export class AnnotationResizeItem extends ToolboxItem {
...
+            front_id_dialog_id, width, this.ulabel.subtasks[current_subtask_key].class_ids,
+            inner_radius, color_info
+        );
+
+        // Remove the old pies
+        id_dialog_container.remove()
+        front_id_dialog_container.remove()
+
+        // Add dialog to the document inside their containers
+        front_subtask_dialog_container_jq.append(front_dialog_html_v2); // TODO(new3d) MOVE THIS TO GLOB BOX -- superimpose atop thee anchor already there when needed, no remove and add back
```
Comment (TrevorBurgoyne) [inline src/toolbox.ts:1426]:
```suggestion
        front_subtask_dialog_container_jq.append(front_dialog_html_v2);
```
Same thing here. This code is mostly duplicated, no? So potentially could be a single function call in both spots

File: src/toolbox.ts [hunk 6]
```diff
@@ -1294,82 +1327,130 @@ export class AnnotationResizeItem extends ToolboxItem {
...
+        subtask_dialog_container_jq.append(dialog_html_v2);
+
+        // Re-add the event listener for changing the opacity on hover
+        let that = this
```
Comment (TrevorBurgoyne) [inline src/toolbox.ts:1430]: ?

File: src/toolbox.ts [hunk 7]
```diff
@@ -1294,82 +1327,130 @@ export class AnnotationResizeItem extends ToolboxItem {
...
+        $(".id_dialog").on("mousemove", function (mouse_event) {
+            if (!that.ulabel.subtasks[current_subtask_key].state.idd_thumbnail) {
+                that.ulabel.handle_id_dialog_hover(mouse_event);
+            }
```
Comment (TrevorBurgoyne) [inline src/toolbox.ts:1434]: Ah, this was what we were missing i guess

File: src/html_builder.ts (hunk 2 above)
Comment (csolbs24) [reply src/html_builder.ts]: [82ac024](https://github.com/SenteraLLC/ulabel/pull/127/commits/82ac0241d5200a8447103f516dabd17c573c5479)

File: src/html_builder.ts (hunk 1 above)
Comment (csolbs24) [reply src/html_builder.ts:28]: Yeah, `md` stands for mode. Definitely not a bad idea to add more function descriptions.

File: src/html_builder.ts (hunk 3 above)
Comment (csolbs24) [reply src/html_builder.ts:386]: No, good catch

File: src/html_builder.ts (hunk 4 above)
Comment (csolbs24) [reply src/html_builder.ts:441]: In my defense this was already present. I believe I just changed the indenting causing it to appear as being added.

File: src/toolbox.ts (hunk 6 above)
Comment (csolbs24) [reply src/toolbox.ts:1430]: I'm not exactly sure what you're questioning? Could you please clarify?

File: src/html_builder.ts (hunk 1 above)
Comment (csolbs24) [reply src/html_builder.ts]: Added function descriptions [1b40b63](https://github.com/SenteraLLC/ulabel/pull/127/commits/1b40b6346fb434a25ac0da280be20c559525b821)

File: src/toolbox.ts (hunk 5 above)
Comment (csolbs24) [reply src/toolbox.ts:1426]: It's only two duplicated lines, so I don't think its worth making a function for it.

File: src/toolbox.ts (hunk 6 above)
Comment (TrevorBurgoyne) [reply src/toolbox.ts:1430]: is there a benefit to making a variable for `that` when all it does is reference `this`?

Comment (TrevorBurgoyne) [reply src/toolbox.ts:1430]: gotcha, that makes sense. Maybe add a note saying you're using `that` bc `this` is a reserved keyword in the event listener

**Review by TrevorBurgoyne — COMMENTED** [review]
Does this also address #96, or is that an unrelated issue?

**Comment by csolbs24** [comment]
> Does this also address #96, or is that an unrelated issue?

No that is a separate issue.

**Comment by csolbs24** [comment]
> Maybe add a note saying you're using that bc this is a reserved keyword in the event listener

Done in [f480cf0](https://github.com/SenteraLLC/ulabel/pull/127/commits/f480cf048da1978532c33791ef39afb178dd6054)

**Review by TrevorBurgoyne — APPROVED** [review]
Nice 👍

**Pull request merged**